### ISSUE TITLE
Discussion: separated checking of variables and types in no-shadowed-variable

### ISSUE THREAD

**Issue by ajafff** [palantir/tslint#3021]
### Bug Report

- __TSLint version__: 5.5.0
- __TypeScript version__:
- __Running TSLint via__: (pick one) CLI / Node.js API / VSCode / grunt-tslint / Atom / Visual Studio / etc

#### TypeScript code being linted

```ts
namespace A {
    interface A {} // error: shadowed name 'A'
    let v = A; // namespace A
    let t: A; // interface A
}
```

#### Actual behavior

Error for shadowed name 'A'

#### Expected behavior

Technically there's no shadowing, because both names can coexist in different domains.
Do we want to allow this at all? By default or with an option?
The implementation would not be too difficult.

**Comment by adidahiya** [comment]
eh, I guess you're right, we could make this change to be technically correct. I don't see this "shadowing" pattern being very valuable in practice so I would prefer it to be opt-in by default (keep the rule strict by default).

**Comment by awk** [comment]
I just updated to tslint 5.6.0 from 4.0 - I've previously had 'shadowed variables' checks enabled however they're now failing for the following pattern of code:

```
  constructor(private $i18next: ngI18Next.I18NextService,
              private $scope: angular.IScope,
              private $uibModal: angular.ui.bootstrap.IModalService,
              private AuthenticationService: AuthenticationService)
```

The use of the type and parameter as the same name (AuthenticationService - this is angular 1.x code) cause the message "Shadowed name: 'AuthenticationService'"

I think it's the same issue as this bug? I tried adjusting the options for no-shadow-variable but adding `"typeAlias": false` && `"typeParameter": false` didn't seem to make any difference?

**Comment by mchambaud** [comment]
Is there a solution to this issue? Other than going back to 4.0?

**Comment by ajafff** [comment]
@mahmoudm The quick fix is to disable checking of everything but parameters and variables to (almost) restore the behavior of 4.x

```js
"no-shadowed-variable": [
  true,
  {
    "class": false,
    "enum": false,
    "function": false,
    "interface": false,
    "namespace": false,
    "typeAlias": false,
    "typeParameter": false,
    "import": false
  }
]
```

**Comment by mchambaud** [comment]
Thanks @ajafff works like a charm!

**Comment by ethanresnick** [comment]
> The use of the type and parameter as the same name (AuthenticationService - this is angular 1.x code) cause the message "Shadowed name: 'AuthenticationService'"

I'd love to see this addressed too. `let SomeModel: SomeModel` or similar shouldn't fail. There's no shadowing or reader ambiguity, and it makes sense for the type and the variable to have the same name in some cases.

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
